lib/GADS/Schema/Result/Report.pm

```diff
@@ -0,0 +1,448 @@
+    required => 0,
+);
+
+=head2 Record ID
+
+This is the ID of the Record in the Instance to display the report for
+
+=cut
+
+has record_id => (
+    is       => 'rw',
+    required => 0,
+);
+
+=head2 Data
+
+This is the data for the report as pulled from the instance record identfied by the Record Id
+
+=cut
+
+has data => (
+    is      => 'rwp',
+    lazy    => 1,
+    builder => sub {
+        my $self = shift;
+
+        my $result = [];
+
+        my $layouts = $self->report_layouts;
+
+        while ( my $layout = $layouts->next ) {
+            my $data = $self->_load_record_data($layout);
+            push( @{$result}, $data );
+        }
+
+        return $result;
+    },
+);
+
+#helper function to load record data
+sub _load_record_data {
+    my $self      = shift;
+    my $layout    = shift;
+    my $record_id = $self->record_id;
+    my $user      = $self->user;
+
+    error "No report id given" if !$record_id;
+
+    my $gads_layout = GADS::Layout->new(
+        schema      => $self->schema,
+        user        => $user,
+        instance_id => $self->instance_id,
+    );
+
+    my $record = GADS::Record->new(
+        schema => $self->schema,
+        user   => $user,
+        layout => $gads_layout,
+    );
+
+    $record->find_current_id($record_id);
+
+    my $column =
+      $self->_find_column( $layout->layout->name, $gads_layout->columns );
+
+    my $datum = $record->get_field_value($column);
+
+    return { 'name' => $layout->layout->name, 'value' => $datum };
+}
+
+#helper function to find a column in a list of columns
+sub _find_column {
+    my $self        = shift;
+    my $column_name = shift;
+    my $columns     = shift;
+
+#I know I could probably do this with a grep, but for some reason, I can't get said grep to work
+#grep { $_->name eq $column_name } @{$columns};
+    foreach my $col ( @{$columns} ) {
+        if ( $col->name eq $column_name ) {
+            return $col;
+        }
+    }
+
+    return undef;
+}
+
+=head1 Object functions
+
+=head2 Update Report
+
+Function to update a report - it requires the schema and any updated fields to be passed in and will return a report object
+
+=cut
+
+sub update_report {
+    my ( $self, $args ) = @_;
+
+    my $guard = $self->schema->txn_scope_guard;
+
+    $self->update( { name => $args->{name} } )
+      if $args->{name} && $args->{name} ne $self->name;
+    $self->update( { description => $args->{description} } )
+      if $args->{description} && $args->{description} ne $self->description;
+
+    my $layouts        = $self->report_layouts;
+    my $report_layouts = [];
+
+    while ( my $layout = $layouts->next ) {
+        push( @{$report_layouts}, $layout->layout_id );
+    }
+
+    #we grep for less writes
+    foreach my $layout (@$report_layouts) {
+        $self->schema->resultset('ReportLayout')
+          ->find( { report_id => $self->id, layout_id => $layout } )->delete
+          if !grep { $_ == $layout } @{ $args->{layouts} };
+    }
```

Comment:
Better to avoid use of grep as it can be quite slow. A better methodology would be to find_or_create (or use a hash), and then use a single DBIx::Class query to do a search of any IDs no longer needed and then to a delete on that. An example is here: https://github.com/ctrlo/GADS/blob/ecc499f05774aedb42d32237bd0edf2b0f23fe4e/lib/GADS/Schema/Result/User.pm#L742